### ISSUE TITLE
`prefer-named-capture-group` should be disabled when using `.test()` or `.replace()`

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#12326]
**What rule do you want to change?**

`prefer-named-capture-group` suggested in https://github.com/eslint/eslint/issues/11381

**Does this change cause the rule to produce more or fewer warnings?**

Fewer warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior

**Please provide some example code that this change will affect:**

```js
const hasPlanet = /with (mars|pluto)/i.test(object);
```
```js
return string.replace(/on (mars|venus)/gi, 'on Pluto');
```

**What does the rule currently do for this code?**

The rule suggests adding `?:`

This has no advantages:
- the result is unchanged
- the group isn't meant to be captured at all
- it lengthens the regex

`.replace` can capture the groups but it doesn't support named groups.

**What will the rule do after it's changed?**

Not report an issue **when** a regex is used inline with `.test` or `.replace()`

**Comment by g-plane** [comment]
Thanks for this issue.

Personally, I think this is expected. You're using regular expressions groups, but you aren't going to capture anything, so this rule suggests you adding prefix `?:` which means using groups without capturing.

**Comment by fregante** [comment]
On the other hand `.replace` needs capture groups **but** it can't use named groups because both the string replacement (`"$1"`) and the replacer function are index-based.

**Comment by mdjermanovic** [comment]
I think that `String.prototype.replace` supports named capture groups

[https://github.com/tc39/proposal-regexp-named-groups#replacement-targets](https://github.com/tc39/proposal-regexp-named-groups#replacement-targets)

**Comment by platinumazure** [comment]
I'm worried about the implementation complexity required to make this change, without type support. The only cases we can catch with minimal complexity would be regex literals that are immediately followed by calls to `.test()` or `.replace()`. Maybe I could be persuaded otherwise, but I'm currently not sure this change is worth doing.

**Comment by ljharb** [comment]
Given that the correct fix for `test` is to add `?:`, and `replace` in the OP should be using `?:`, and also supports named capture groups - it seems like there's nothing that needs changing in the first place?

**Comment by fregante** [comment]
> it seems like there's nothing that needs changing in the first place?

In the case of `.test()` it doesn't _need_ changing, but also `?:` are just extra characters that don't benefit anyone since the result is a boolean.

**Comment by ljharb** [comment]
Sure, but does adding it potentially make your regex more performant, since the engine doesn’t have to capture the group and look for backreferences?
